Source/AwsCommonRuntimeKit/io/HostResolver.swift

```diff
@@ -59,15 +78,60 @@ public class HostResolver {
                     hostResolutionConfigPointer,
                     continuationCore.passRetained()) != AWS_OP_SUCCESS {
 
-                    // TODO: this is wrong. Sometimes it triggers the error callback and sometimes it doesn't.
-                    // I have a fix in progress in aws-c-io
                     continuationCore.release()
                     continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
                 }
             }
         })
     }
 
+    /// Reports a failure on an address so that the background cache can accommodate
+    /// the failure and likely not return the address until it recovers.
+    /// Note: While the underlying C API may report an error, we ignore it because users don't care and
+    /// don't have a good way to deal with it.
+    public func reportFailureOnAddress(address: HostAddress) {
+        address.withCPointer { cAddress in
+            _ = aws_host_resolver_record_connection_failure(rawValue, cAddress)
+        }
+    }
+
+    /// Purges the cache for a specific address
+    /// Note: While the underlying C API may report an error, we ignore it because users don't care and
+    /// don't have a good way to deal with it.
+    public func purgeCache(args: HostResolverArguments) async {
+        let host = AWSString(args.hostName, allocator: allocator)
+        return await withCheckedContinuation({ (continuation: CheckedContinuation<(), Never>) in
+            let continuationCore = Box(continuation)
+            var purgeCacheOptions = aws_host_resolver_purge_host_options()
+            purgeCacheOptions.host = UnsafePointer(host.rawValue)
+            purgeCacheOptions.on_host_purge_complete_callback = onPurgeCacheComplete
+            purgeCacheOptions.user_data = continuationCore.passRetained()
+            guard aws_host_resolver_purge_host_cache(rawValue, &purgeCacheOptions) == AWS_OP_SUCCESS else {
+                continuationCore.release()
+                continuation.resume()
+                return
+            }
```

Comment:
if guard's else is not executed, we are not resuming continuation. There must be at least one resume.

Reply:
It is in the callback.